File: 39-CombinationSum/39-CombinationSum.java
// Last updated: 7/2/2025, 11:55:21 PM
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        help(candidates , target , res , new ArrayList<>(),  0);
        return res;
    }
    public void help(int[] arr , int tar, List<List<Integer>> l, ArrayList<Integer> lcl , int i)
    {
        if(i == arr.length || tar<=0)
        {
            if(tar == 0) l.add(new ArrayList<>(lcl));
            return;
        }
        lcl.add(arr[i]);
        help(arr , tar-arr[i], l, lcl, i);
        lcl.remove(lcl.size()-1);
        help(arr, tar , l, lcl , i+1);
    }
}

File: 39-CombinationSum/39-CombinationSum.cpp
// Last updated: 7/2/2025, 11:55:17 PM
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {

        vector<vector<int>> ans;
        vector<int> lcl;

        help(candidates, 0,target, ans, lcl);
        
        return ans;
        
    }

    void help(vector<int>& arr, int idx, int tar, vector<vector<int>> &ans,
    vector<int> &lcl)
    {
        if(tar == 0)
        {
            ans.push_back(lcl);
            return;
        }

        for(int i = idx ; i < arr.size();i++)
        {
            if(tar - arr[i] >= 0)
            {
                lcl.push_back(arr[i]);
                help(arr,i,tar-arr[i],ans,lcl);
                lcl.pop_back();
            }
        }
    }
};